build: download wintools executables separately

dundargoc opened the pull request:
The wintools executables are stored in a zip file, making it
inconvenient to bump these during releases. Instead, unpack the
executables in the deps repository and download each executable
separately.

dundargoc committed:
build: download wintools executables separately

The wintools executables are stored in a zip file, making it
inconvenient to bump these during releases. Instead, unpack the
executables in the deps repository and download each executable
separately.

dundargoc commented:
@justinmk very rough draft, something like this should work. `deps` will have the following structure: https://github.com/dundargoc/deps/tree/master/opt

Will need to clean this up, this is the best I could come up with due to potato laptop limitations.

dundargoc commented:
Actually unsure if this is even needed. We should be able to download the entire repo as an archive from github directly.

justinmk commented on cmake.deps/CMakeLists.txt:
```diff
@@ -162,9 +162,29 @@ set(LUV_SHA256 16720d17b9e6d42a7008b902207ea2d61e3eeaef3cf358398dbbf3777867ec09)
-set(WINTOOLS_SHA256 3c4c490a3d392ceeb1347cb77cc821a31900b688a2189276d3a1131a3f21daf1)
+set(CAT_URL https://github.com/dundargoc/deps/raw/master/opt/cat.exe)
+set(CAT_SHA256 93b8d307bb15af3968920bdea3beb869a49d166f9164853c58a4e6ffdcae61c6)
+
```
nit: let's group these all together (no blank lines) and add a comment like `// Windows only: curl.exe, xxd.exe, etc.`

justinmk commented on cmake.deps/CMakeLists.txt:
```diff
@@ -263,8 +283,93 @@ endif()
+    CONFIGURE_COMMAND ""
+    BUILD_COMMAND ""
+    INSTALL_COMMAND ${CMAKE_COMMAND} -E make_directory ${DEPS_BIN_DIR}
+    COMMAND ${CMAKE_COMMAND} -E copy <DOWNLOADED_FILE> ${DEPS_BIN_DIR})
```
would `GetBinaryDep` avoid some boilerplate here?

dundargoc replied:
A variation of it, but in hindsight I don't think it's needed. A single download should be sufficient. I will investigate properly when I can.

dundargoc replied:
Fixed

justinmk commented:
>  unsure if this is even needed. We should be able to download the entire repo as an archive from github directly.

Could, though might be slow unless we do shallow/partial clone. Also, direct URLs makes these cases a bit less "special" I think?

dundargoc commented:
> Could, though might be slow unless we do shallow/partial clone.

No, I meant the "archive" feature github has, to download any commit/tag as an archive like we do with some of our other dependencies like unibilium.

> Also, direct URLs makes these cases a bit less "special" I think?

Can you elaborate, missed what you mean.

dundargoc commented:
The unstable PPA builds requires the `deps` repo to be as is, but it's probably too big to download the entire repo. Going back to the original plan of downloading required executables one by one.

dundargoc commented:
@justinmk what about this? If it looks alright I can send PR to the deps repo.

justinmk commented on cmake.deps/CMakeLists.txt:
```diff
@@ -163,9 +163,19 @@ set(LUV_SHA256 16720d17b9e6d42a7008b902207ea2d61e3eeaef3cf358398dbbf3777867ec09)
+set(CURL_CA_BUNDLE_URL https://github.com/dundargoc/deps/raw/master/opt/curl-ca-bundle.crt)
+set(CURL_CA_BUNDLE_SHA256 6ed95025fba2aef0ce7b647607225745624497f876d74ef6ec22b26e73e9de77)
+set(CURL_URL https://github.com/dundargoc/deps/raw/master/opt/curl.exe)
+set(CURL_SHA256 4c7151dea34c12b351030d4a14e50336a5a4e4639726565829997eea0d744654)
```
I vaguely recall curl involving several files. I assume you checked that `curl` works on windows with only these 2 files?

justinmk commented on cmake.deps/cmake/GetBinaryDeps.cmake:
```diff
@@ -10,17 +10,10 @@ function(GetBinaryDep)
   set(URL ${${URL_VARNAME}})
   set(HASH ${${HASH_VARNAME}})
-  if(NOT URL OR NOT HASH )
-    message(FATAL_ERROR "${URL_VARNAME} and ${HASH_VARNAME} must be set")
```
Why isn't this needed/wanted?

dundargoc replied:
ExternalProject will emit the same warning anyway, so feels a bit unnecessary.

justinmk reviewed (approved):
nice

dundargoc commented:
TODO: don't merge yet, need to change URL to the actual neovim dep url. Meed ro change the deps repo first.

dundargoc commented:
> Windows 10 comes with the curl tool bundled with the operating system since version 1804.

Drop curl from neovim?

dundargoc commented:
> Windows 10 comes with the curl tool bundled with the operating system since version 1804.

@justinmk thoughts?

dundargoc committed:
build: drop curl.exe on Windows

Curl is already shipped by default on Windows 10 starting from versions
1803. As we already require version 1809 we can safely remove it.

justinmk reviewed (approved):
🚀

clason commented:
@dundargoc are these reasonably up-to-date version, or should we bump them?

dundargoc commented:
xxd is "version" 2022-01-14
cat is v2.1, latest is v9.2
tee doesn't have a version, it just says in the version description that "this version of tee is supplied with Vim 4.2 to make ':make' possible"
diff is v2.7 (Oct-1994), latest is 3.9
